### DISCUSSION TITLE
Duplicate ID error when creating a push with the same ID after stopping a push

### DISCUSSION THREAD

**Discussion by vampirefrog** [OvenMediaLabs/OvenMediaEngine#1556] [Ideas]
**Describe the bug**
When I try to call `:startPush` after calling `:stopPush` I get the error `Duplicate ID (400)` if the stopped push and the started push have the same ID.

**To Reproduce**
```
POST> http://127.0.0.1:60080/v1/vhosts/default/apps/live:stopPush? {"id":"ZZZZZ"} {"Content-type":"application/json","authorization":"Basic XXXX"}
{"message":"OK","statusCode":200}

POST> http://127.0.0.1:60080/v1/vhosts/default/apps/live:startPush? {"id":"ZZZZZ","stream":{"name":"WWWWW","variantNames":["bypass_video","bypass_audio"]},"protocol":"rtmp","url":"rtmps://xxxxx:443/rtmp","streamKey":"YYYYYY"} {"Content-type":"application/json","authorization":"Basic XXXX"}
{"message":"[HTTP] Duplicate ID (400)","statusCode":400}
```
In this case the ID "ZZZZZ" is the same for both requests. This issue seems to occur both when the stream (`WWWWW` in this example) is receiving RTMP and when it's not receiving.

**Expected behavior**
I'd expect the result to be 200, not 400.

**Logs**
```
[2024-03-15 09:25:41.913] I [SPLLHLS-t60080:2746237] HTTP.Server | http_server.cpp:149  | Client(<ClientSocket: 0x7ff146c07210, #14, Connected, TCP, Nonblocking, 127.0.0.1:35020>) is connected on *:60080
[2024-03-15 09:25:41.916] I [SPLLHLS-t60080:2746237] HTTP.Server | http_server.cpp:149  | Client(<ClientSocket: 0x7ff146c07a10, #16, Connected, TCP, Nonblocking, 127.0.0.1:35026>) is connected on *:60080
[2024-03-15 09:25:41.919] W [SPLLHLS-t60080:2746237] APIController | controller.h:154  | HTTP error occurred: [HTTP] Duplicate ID (400)
[2024-03-15 09:25:41.919] E [SPLLHLS-t60080:2746237] HTTP.Server | http_exchange.cpp:42   | 
[Client] HttpConnection(0x7ff146c07c10) : HTTP/1.1 <ClientSocket: 0x7ff146c07210, #14, Connected, TCP, Nonblocking, 127.0.0.1:35020> TLS(Disabled) (Elapsed: 0) 
[Request] POST http://127.0.0.1:60080/v1/vhosts/default/apps/live:startPush (HTTP/1.1, Request Time: 2024-03-15T09:25:41.919+02:00)
[Response] 400 Bad Request (364 bytes sent, Response Time: 2024-03-15T09:25:41.919+02:00)
[2024-03-15 09:25:46.917] I [SPLLHLS-t60080:2746237] HTTP.Server | http_server.cpp:223  | Client(<ClientSocket: 0x7ff146c07210, #14, Disconnected, TCP, Nonblocking, 127.0.0.1:35020>) has disconnected from *:60080
[2024-03-15 09:25:46.918] I [SPLLHLS-t60080:2746237] HTTP.Server | http_server.cpp:223  | Client(<ClientSocket: 0x7ff146c07a10, #16, Disconnected, TCP, Nonblocking, 127.0.0.1:35026>) has disconnected from *:60080
```

**Server (please complete the following information):**
 - OS: Debian 12.5
 - OvenMediaEngine Version: 1.16.4 
 - Branch: ((From archive))

**Player (please complete the following information):**
N/A (bug is server side)

**Additional context**
I've been aware of this bug for a while, but I didn't think it would pose an issue. Basically when you call :stopPush, the push still shows up in the list of pushes. In my case I wanted to change the RTMP URL to a closer ingest and that's when this issue happened. I believe the stopped push does go away if the stream stops, but in my case I couldn't stop the stream.

**Comment by getroot** [comment]
Is there any reason to use the same ID again? This is by design. Because IDs can also be used to look up past records, IDs that have been used once cannot be used.

**Comment by vampirefrog** [comment]
I am using the ID to match my app's internal IDs, the format is `<OME server ID>-<Service ID>-<target ID>`, so for example `1-1-vampirefrog`. This way I don't have to store a randomly generated ID.

**Comment by vampirefrog** [comment]
I'm not sure how keeping an ID around makes sense for pushes, it only makes sense for recordings.

**Comment by vampirefrog** [comment]
Also, if you're keeping it, how come it clears when a stream goes online or offline, and doesn't stay around indefinitely?

**Comment by getroot** [comment]
We may push and stop the stream multiple times while it is online, which can later be tracked by ID in the log. There are various cases. Well... what you need might be deletePush or resumePush. I'll consider this.

**Reply by vampirefrog** [reply]
Ah, interesting, I hadn't considered that. For my use case, I'll just add a random suffix which I will later ignore, thanks for the clarification.

**Reply by bchah** [reply]
The way I implement it is:

- User defines a "restream" (our term for push) with a push URL and we assign it a static ID
- User starts the "restream" and we create a push on OME with a random ID, stored as a value "current_push"
- We internally query OME with the id of 'current_push' for status updates or control
- User stops the "restream" and we internally stop the push. Next time the push will be created with a new random ID, and so on...

Good luck in your implementation!

**Reply by vampirefrog** [reply]
> We may push and stop the stream multiple times while it is online, which can later be tracked by ID in the log. There are various cases. Well... what you need might be deletePush or resumePush. I'll consider this.


By this logic, wouldn't you also need a "createPush" that creates a push that can be started later?

So, so far we have `createPush`, `deletePush`, `startPush`, `stopPush`. With `createPush` you can have a boolean flag `start` that starts it when you create it, so you don't have to call `startPush` as well. I don't think you need `resumePush`

**Comment by getroot** [comment]
This is not a bug, so moving on to discussion.
